### ISSUE TITLE
custom attention vs allennlp attention

### ISSUE THREAD

**Issue by antgr** [allenai/allennlp#3457]
Hi,
I have implemented the following attention function, adapting it from an, as it called "attention bi lstm" class I found,

def attention(out, hidden):
    out = out.permute(1,0,2)
    hidden = hidden.squeeze(0)
    attn_weights = torch.einsum('pqr,pr->pq', [out, hidden])
    soft_attn_weights = F.softmax(attn_weights, 1)
    new_hid = torch.einsum('pqr,pq->pr', [out, soft_attn_weights])
    return new_hid

as I mentioned, this is intended to be used after an lstm, and works in my code.

Calling it with

out: torch.Size([2, 2, 3])
hidden: torch.Size([2, 3])

gives torch.Size([2, 3])

But if I use allennlp attention implementations, I have:

linear = CosineAttention(normalize=False)
output = linear(
            torch.FloatTensor([[0, 0, 0], [1, 1, 1]]),
            torch.FloatTensor([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]]),
        )
output.shape
torch.Size([2, 2])

[2, 3], [2, 2, 3] ---> [2, 2]

linear = DotProductAttention(normalize=False)
output = linear(
  torch.FloatTensor([[0, 0, 0], [1, 1, 1]]),
  torch.FloatTensor([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]]),
)
output.shape
torch.Size([2, 2])

[2, 3], [2, 2, 3] ---> [2, 2]

Can you provide some feedback? Is my implementation wrong?
I would like to replace my implementation with allennlp's attention functions!
But is those implementations appropriate for applying them after an lstm output?
Your help would be much appreciated.

**Comment by antgr** [comment]
I forgot: https://discourse.allennlp.org/
now my question posted there.

**Comment by antgr** [comment]
I forgot: https://discourse.allennlp.org/
now my question posted there.

**Comment by matt-gardner** [comment]
I responded [on discourse](https://discourse.allennlp.org/t/custom-attention-vs-allennlp-attention/131/2).  Closing this issue as it's a question, not a bug or feature request, and it's duplicated on discourse.
